kotlinx-coroutines-core/jvm/src/scheduling/parallelismCompensation.kt

```diff
@@ -11,21 +26,86 @@ private val parallelismCompensationEnabled: Boolean =
  * that eventually it will adjust to meet it.
  */
 @Suppress("NOTHING_TO_INLINE") // better stacktrace
-internal inline fun <T> withCompensatedParallelism(noinline body: () -> T): T {
+internal inline fun <T> withCompensatedParallelism(timeout: Duration, noinline body: () -> T): T {
     if (!parallelismCompensationEnabled) {
         return body()
     }
     // CoroutineScheduler.Worker implements ParallelismCompensation
     val worker = Thread.currentThread() as? ParallelismCompensation
         ?: return body()
-    return worker.withCompensatedParallelism(body)
+    return if (timeout == Duration.ZERO) {
+        worker.withCompensatedParallelismImmediate(body)
+    } else {
+        worker.withCompensatedParallelismAfterDeadline(timeout, body)
+    }
 }
 
-private fun <T> ParallelismCompensation.withCompensatedParallelism(body: () -> T): T {
+private fun <T> ParallelismCompensation.withCompensatedParallelismImmediate(body: () -> T): T {
     increaseParallelismAndLimit()
     try {
         return body()
     } finally {
         decreaseParallelismLimit()
     }
+}
+
+private fun <T> ParallelismCompensation.withCompensatedParallelismAfterDeadline(timeout: Duration, body: () -> T): T {
+    val holder = AtomicValueHolder()
+    val compensationJob = initCompensationWithDeadline(this, holder, timeout)
+    try {
+        return body()
+    } finally {
+        endCompensationWithDeadline(this, holder, compensationJob)
+    }
+}
+
+private class AtomicValueHolder() {
+
+    private enum class ParallelismCompensationState {
+        NONE,
+        PARALLELISM_INCREASED,
+        TASK_FINISHED
+    }
+
+    private val _value: AtomicRef<ParallelismCompensationState> = atomic(ParallelismCompensationState.NONE)
+    fun tryCompleteWithCompensation(): Boolean {
+        return _value.compareAndSet(ParallelismCompensationState.NONE, ParallelismCompensationState.PARALLELISM_INCREASED)
+    }
+
+    fun tryCompleteWithTask(): Boolean {
+        return _value.compareAndSet(ParallelismCompensationState.NONE, ParallelismCompensationState.TASK_FINISHED)
+    }
+}
+
+private fun initCompensationWithDeadline(
+    thread: ParallelismCompensation,
+    state: AtomicValueHolder,
+    timeout: Duration
+): Job {
+    // By default, `GlobalScope` launches coroutines on `Dispatchers.Default`.
+    // This is definitely not what we want, as our aim is to be able to wake up the default dispatcher.
+    return GlobalScope.launch(isolatedWakingDispatcher, start = CoroutineStart.UNDISPATCHED) { // start waiting as early as possible
+        delay(timeout)
+        if (state.tryCompleteWithCompensation()) {
+            // The deadline for compensation arrived, now we need to unblock the thread
+            thread.increaseParallelismAndLimit()
```

Comment:
Unfortunately, the implementation of this method for Worker relies on the fact that it is called on the same thread only. For example, it consults the thread's state (do we have a CPU permit or not). Running this code from a different thread leaves a window for a race condition at the very least